openscope/openscope
# devops/1749 - More robust unit tests for input validation, improved validation code

Resolves #1749 

The purpose of this pull request is to make unit tests for input validation more robust and improve validation code accordingly

### Scope
This covers all methods that have validation involving one or more of `_isEmpty`, `_isArray`, `_isObject` either directly or via the helper methods in `validatorUtilities.js`. This generally means constructors and `init` methods, usually dealing with data ingested from json files.

Was originally intended to cover anything that `throw new TypeError` but that turned out to be too much to swallow at once. 

### Validation code changes
- where constructor or `init` takes multiple arguments, check for missing args first before validating individual args
  - this facilitates writing tests that expect certain error messages; otherwise when there are missing args, which error message you get is dependent on the order in which args get validated.
- use `isEmptyOrNotArray` or `isEmptyOrNotObject` helpers as appropriate
- use `instanceof` checks as appropriate
- use a consistent error message wording throughout
- drop pointless `else`

### Tests changes
- tests `t.throws()` now check against expected error message; this ensure throws are indeed from the validation being tested and not due to other reasons
- separate test for missing args when multiple arguments are involved
- add tests for new validation code
- add missing tests

### Misc. changes
- reordered tests in `Fms.spec.js` to be in same order that the respective tested methods appear in `Fms.js`
  - originally intended to make more changes; soon after this I realized I needed to rein in the scope of the PR so those changes were dropped, but no harm keeping the reordering.
- in `Fms.js` change [this `throws`](https://github.com/openscope/openscope/blob/bc3caf812627b6c073eed7b96d6c4c088abb7b3d/src/assets/scripts/client/aircraft/FlightManagementSystem/Fms.js#L936) in `replaceDepartureProcedure` to a `return [false, <readback>]`
  - makes consistent with the rest of the `replaceDepartureProcedure` method
  - originates from user command, should be displayed to user as readback in log rather than thrown to console
  - in #1002, was originally a `return` before refactoring in commit 491573e (inadvertently?)changed it to a `throw`
- discard defunct `isEmptyObject` from validation helpers
- typo fixes
